### PR TITLE
feat(database) migrate database(postgresql) LCM controller

### LINKED ISSUE #150: Migrate database(postgresql) LCM controller

### PR DESCRIPTION
**Pull request by Ghostbaby**
Migrate database(postgresql) LCM controller

Fix #150

Signed-off-by: zhuhuijun <zhuhuijunzhj@gmail.com>

### CODE REVIEW

File: controllers/goharbor/harborcluster/database/api/postgresql_type.go [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+
+import (
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// Postgresql defines PostgreSQL Custom Resource Definition Object.
+type Postgresql struct {
```
Comment (wangcanfengxs) [inline controllers/goharbor/harborcluster/database/api/postgresql_type.go:13]: Why need to rewrite the Postgresql CRD definition? Package import is not worked?

File: pkg/k8s/option.go [hunk 2]
```diff
@@ -0,0 +1,16 @@
+package k8s
+
+import (
+	"context"
+
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+type GetOptions struct {
```
Comment (wangcanfengxs) [inline pkg/k8s/option.go:10]: I think we should refactor the struct name？GetOptions is not optional, and there is used to create a controller, not get a controller.

File: controllers/goharbor/harborcluster/database/postgresql.go [hunk 3]
```diff
@@ -1,15 +1,74 @@
...
 }
 
 func (p *PostgreSQLController) Apply(harborcluster *v1alpha2.HarborCluster) (*lcm.CRStatus, error) {
-	panic("implement me")
+
+	p.Client.WithContext(p.Ctx)
+	p.DClient.WithContext(p.Ctx)
+	p.HarborCluster = harborcluster
+
+	crdClient := p.DClient.WithResource(databaseGVR).WithNamespace(p.HarborCluster.Namespace)
```
Comment (wangcanfengxs) [inline controllers/goharbor/harborcluster/database/postgresql.go:45]: How does it work when using external database

**Review by wangcanfengxs — COMMENTED** [review]
comments left

File: controllers/goharbor/harborcluster/database/postgresql.go (hunk 3 above)
Comment (soulseen) [reply controllers/goharbor/harborcluster/database/postgresql.go:45]: If an external database is used, will this function be executed?

File: controllers/goharbor/harborcluster/database/postgresql.go [hunk 4]
```diff
@@ -1,15 +1,74 @@
 package database
```
Comment (soulseen) [inline controllers/goharbor/harborcluster/database/postgresql.go:1]: lcm controlles please move to dir `/pkg/cluster`

File: controllers/goharbor/harborcluster/database/postgresql.go [hunk 5]
```diff
@@ -20,6 +79,12 @@ func (p *PostgreSQLController) Upgrade(harborcluster *v1alpha2.HarborCluster) (*
 	panic("implement me")
 }
 
-func NewDatabaseController() lcm.Controller {
-	return &PostgreSQLController{}
+func NewDatabaseController(options *k8s.GetOptions) lcm.Controller {
```
Comment (steven-zou) [inline controllers/goharbor/harborcluster/database/postgresql.go:82]:
```suggestion
func NewDatabaseController(options *k8s.CtrlOptions) lcm.Controller {
```

File: controllers/goharbor/harborcluster/harborcluster_controller.go [hunk 6]
```diff
@@ -35,20 +43,45 @@ func (r *Reconciler) SetupWithManager(ctx context.Context, mgr ctrl.Manager) err
...
+		return nil, err
+	}
+
+	client, err := k8s.NewClient(scheme)
+	if err != nil {
+		return nil, err
+	}
+
+	option := &k8s.GetOptions{
+		CXT:     ctx,
```
Comment (steven-zou) [inline controllers/goharbor/harborcluster/harborcluster_controller.go:58]:
```suggestion
		Context:     ctx,
```

File: controllers/goharbor/harborcluster/database/api/postgresql_type.go (hunk 1 above)
Comment (Ghostbaby) [reply controllers/goharbor/harborcluster/database/api/postgresql_type.go:13]: yes, The introduction of official CRD will cause version conflicts.

File: pkg/k8s/option.go (hunk 2 above)
Comment (Ghostbaby) [reply pkg/k8s/option.go]: yes, I will rename GetOptions to CtrlOptions.

**Review by steven-zou — COMMENTED** [review]
commented

File: controllers/goharbor/harborcluster/database/postgresql.go (hunk 3 above)
Comment (Ghostbaby) [reply controllers/goharbor/harborcluster/database/postgresql.go:45]: Only HarborCluster CR defines InClusterDatabase will execute this func.

File: controllers/goharbor/harborcluster/database/postgresql.go (hunk 4 above)
Comment (Ghostbaby) [reply controllers/goharbor/harborcluster/database/postgresql.go:1]: copy

File: config/samples/harborcluster/goharbor_v1alpha2_harborcluster.yaml [hunk 7]
```diff
@@ -5,6 +5,18 @@ metadata:
 spec:
   harborAdminPasswordRef: admin-core-secret
   externalURL: https://core.harbor.domain
+  redis:
+    host: harbor-redis-master
+    port: 6379
+    passwordRef: harbor-redis
+  database:
```
[the remark this answers is not in the document]
Comment (Ghostbaby) [reply config/samples/harborcluster/goharbor_v1alpha2_harborcluster.yaml:12]: yes, but some of the struct items define as required.

[the remark this answers is not in the document]
Comment (steven-zou) [reply config/samples/harborcluster/goharbor_v1alpha2_harborcluster.yaml:12]: @wangcanfengxs we need to revise the CRD spec?

**Review by steven-zou — APPROVED** [review]
LGTM

**Comment by wangcanfengxs** [comment]
LGTM

**Review by soulseen — COMMENTED** [review]
pls resolve conflicts

**Review by wangcanfengxs — APPROVED** [review]

**Pull request merged**